### PR TITLE
New build flow (promote builds after delay)

### PR DESCRIPTION
**Pull request by gabriel**
https://keybase.atlassian.net/browse/DESKTOP-641

### CODE REVIEW

**Comment by cjb** [comment]
@gabriel please ping some reviewers :)

**Comment by cjb** [comment]
@keybase/react-hackers 

I wonder if we could do e.g.:

Tue 9am ET: master released to admin
Tue 12pm ET: Mon 9am ET build released to public
Wed 9am ET: master released to admin
Wed 12pm ET: Tue 9am ET build released to public

So it's a 27 hour delay from public to admin, and there are admin builds at 9am ET and public builds at 12pm ET.  That way we get 24 hours to test the non-update parts of Build X, and 3 hours to test Build X's capability to update to Build Y, before Build X goes out to the public.

Does that sound too complex?  It solves the problem that the 24 hour delay doesn't actually test the build's ability to perform an update, which should be the *only* thing we care about testing hard during the delay.

File: go/libkb/config.go [hunk 1]
```diff
@@ -672,6 +672,10 @@ func (f JSONConfigFile) GetUpdateURL() string {
 	return s
 }
 
+func (f JSONConfigFile) IsAdmin() (bool, bool) {
```
Comment (MarcoPolo) [inline go/libkb/config.go]: Why does this return 2 bools?

Comment (gabriel) [inline go/libkb/config.go]: Second bool is if a value was present at all (if set)... 

For example you might want the default value to be true (if not set) in which case you check if not set and return true.

I'm following the convention of returning bools this way in this class for consistency.

**Comment by gabriel** [comment]
@cjb that plan sounds good to me

**Comment by gabriel** [comment]
Can we move it to 12pm ET and 3pm ET ... that way it gives west coast people time to see the update and also make sure it's ok

**Comment by cjb** [comment]
9am and 3pm ET OK?

NYC folks like to test first thing, and I'd rather not have changes NUC just pushed to master in there.

**Comment by gabriel** [comment]
Yeah.. ok I think this is ready

The builds scripts here create update json files like:
darwin-support/update-darwin-prod-1.0.15+blah.json

The release tool here does the promoting of releases:
https://github.com/keybase/release/pull/4

It looks for a build older than 27 hours, that was build before 9:15am that day.

Test channel builds are always promoted immediately

**Comment by gabriel** [comment]
cc @oconnor663

File: packaging/desktop/package_darwin.sh [hunk 2]
```diff
@@ -196,9 +197,16 @@ create_zip() {
   ditto -c -k --sequesterRsrc --keepParent $app_name.app $zip_name
 }
 
-s3sync() {
-  if [ ! "$bucket_name" = "" ] && [ ! "$save_dir" = "" ]; then
-    s3cmd sync --acl-public --disable-multipart $save_dir/* s3://$bucket_name/
+kbsign() {
+  cd $out_dir
```
Comment (oconnor663) [inline packaging/desktop/package_darwin.sh]: A `cd` here persists after this function exits -- could be confusing if calling code doesn't realize that the cwd has changed. Might be safer to put parens around these two lines, to contain the effect of the `cd`?

File: packaging/desktop/package_darwin.sh [hunk 3]
```diff
@@ -196,9 +197,16 @@ create_zip() {
...
+  keybase sign -d -i "$zip_name" -o "$sig_name"
+}
+
+update_json() {
+  cd $out_dir
```
Comment (oconnor663) [inline packaging/desktop/package_darwin.sh]: Ditto.

File: packaging/desktop/package_darwin.sh [hunk 4]
```diff
@@ -196,9 +197,16 @@ create_zip() {
...
+  if [ -n "$s3host" ]; then
+    $release_bin update-json --version="$app_version" --src="$zip_name" \
```
Comment (oconnor663) [inline packaging/desktop/package_darwin.sh]: I like to be religious about quoting variables like `$release_bin`.

File: packaging/desktop/package_darwin.sh [hunk 5]
```diff
@@ -206,28 +214,32 @@ save() {
...
+  mkdir -p "$platform_dir-support"
+  mv "$out_dir/$update_json_name" "$platform_dir-support"
+}
 
-  if [ ! "$s3host" = "" ]; then
-    keybase sign -d -i "$platform-updates/$zip_name" -o $sig_path
-    $release_bin update-json --version=$app_version --src="$platform-updates/$zip_name" --uri="$s3host/$platform-updates" --signature="$sig_path" > update-$platform-$run_mode.json
+s3sync() {
+  if [ ! "$bucket_name" = "" ] && [ ! "$save_dir" = "" ]; then
+    s3cmd sync --acl-public --disable-multipart $save_dir/* s3://$bucket_name/
```
Comment (oconnor663) [inline packaging/desktop/package_darwin.sh]: Quoting `$save_dir` is still possible if you leave the `*` outside, something like `"$save_dir"/*`.

**Comment by oconnor663** [comment]
Just some bash nits above.

Talked to @gabriel over hangouts about what to do with the Linux build. We can get a similar effect, but I'd probably need to do it by building a third flavor of the packages, which might take a day or two of work. @cjb, what do you think about the priority of that? Linux might not need the extra testing as badly, but doing very different things on the two platforms could be confusing.

**Comment by cjb** [comment]
I agree it's not very important.  I'd say we can leave Linux as-is for now.

**Pull request merged**